tests/unit/test_adapter.py

```diff
@@ -0,0 +1,236 @@
+#
+# Copyright 2023 SUSE LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from unittest import mock
+
+import pytest
+
+import csp_billing_adapter
+from csp_billing_adapter.adapter import (
+    event_loop_handler,
+    get_config,
+    get_plugin_manager,
+    initial_adapter_setup,
+    main as cba_main,
+    setup_logging
+)
+from csp_billing_adapter.utils import (
+    get_now,
+    string_to_date
+)
+
+
+def test_adapter_version():
+    """Verify the adapter version."""
+    assert csp_billing_adapter.__version__ == "0.0.1"
+
+
+def test_get_plugin_manager(cba_config):
+    """Verify that get_plugin_manager() works correctly."""
+    pm = get_plugin_manager()
+
+    # NOTE: we may need to explicitly pm.register(local_csp) here if
+    # we stop registering it in get_plugin_manager()
+
+    assert pm.hook.get_csp_name(config=cba_config) == "local"
+
+
+def test_setup_logging(cba_pm):
+    """Verify logging is being setup correctly."""
+    log = setup_logging(cba_pm.hook)
+
+    assert log.name == "CSPBillingAdapter"
+
+
+def test_get_config(cba_pm, cba_config, cba_config_path):
+    """Verify correct operation of get_config()."""
+    config = get_config(cba_config_path, cba_pm.hook)
+
+    assert config == cba_config
+
+
+def test_initial_adapter_setup(cba_pm, cba_config):
+    """
+    Verify that the initial_adapter_setup() works correctly using
+    in-memory plugins.
+    """
+    initial_adapter_setup(cba_pm.hook, cba_config)
+
+    cache = cba_pm.hook.get_cache(config=cba_config)
+    assert cache != {}
+
+    csp_config = cba_pm.hook.get_cache(config=cba_config)
+    assert csp_config != {}
+
+
+@mock.patch('csp_billing_adapter.local_csp.randrange')
+def test_event_loop_handler(mock_randrange, cba_pm, cba_config):
+    """Verify correct operation of event_loop_handler()."""
```

Comment:
This is a really long test. Would it be possible to add some more details in the docstring explaining the flow of the 4 different "now" values?

Reply:
Sure.

Reply:
Updated with comments